Enable LTDC/DSI on stm32f469i-disco using PLLSAI

phildefer opened the pull request:
To complement Pull Request #95200, this PR adds configuration of PLLSAI for stm32f469i-disco.
It enables configurations of LTDC & DSI-HOST blocks to use display and touchscreen on this board.
To use framebuffer required for ltdc & dsi drivers, it enables FMC/SDRAM also on this board.

This PR has been successfully tested using samples/drivers/display, samples/subsys/display/lvgl & 
samples/subsys/inputs_draw_events applications

phildefer committed:
dts: arm: st: add mipi_dsi node in stm32f469.dtsi

Describe mipi_dsi block available on stm32f469 & above
Allow to display data on DSI panels taking output of LTDC
after serializing data.

Signed-off-by: Philippe Peurichard <p.peurichard@gmail.com>

phildefer committed:
boards: st: add pllsai for ltdc & mipi-dsi support on stm32f469 disco

Extension of support of pllsai for display configuration of stm32f469
discovery board. Enable Display panel through LTDC & DSI-HOST
blocks. Enable Touch screen. Enable FMC/SDRAM for Framebuffer.

Signed-off-by: Philippe Peurichard <p.peurichard@gmail.com>

phildefer committed:
samples: drivers: display: add stm32f469_disco conf file

Add stm32f749i_disco conf file to support samples display driver
applications . Increase the amount of HEAP to ensure
k_malloc allocation goes well.

Signed-off-by: Philippe Peurichard <p.peurichard@gmail.com>

zhang-wenchao commented:
If possible, please also fix `st_b_lcd40_dsi1_mb1166.overlay`, the I2C reg `0x38` in it is wrong. This is causing confusion for everyone using this mb1166 board.

zhang-wenchao commented:
> Actually, this PR fully depends on #95200. It is what i wanted to mention in the beginning of the description. From today, PR #95200 was reviewed successfully, i hope it will be merged soon. On stm32f469i-disco board, there is no st_b_lcd40_dsi1_mb1166 shield, as such even if same display panel is used.

I misread the board. This is f4. I should reply to another PR, but the other PR is waiting to be merged.

phildefer commented:
@JarmouniA : display is fully integrated onto the stm32f469i-disco and it is not the mb1166 shield as on some other STM32 boards. That is the reason why changes has been made directly in board dts file and not using st_b_lcd40_dsi1_mb1166 overlay.

avolmat-st reviewed (changes requested):
Thanks @phildefer.

Indeed, while this is very close (is it really exactly same ?) to the MB1166 (either normal or a0 variant), this is directly connected to the board via a different 25pins connector different from the usual 30pins DSI/I2C connector.
I have been wondering if it would be possible to directly attach (by default) the shield st_b_lcd40_dsi1_mb1166 to this board (kind of select SHIELD_ST_B_LCD40_DSI1_MB1166 with this board but that doesn't seem possible.
Benefit of it would have been to avoid duplication.

A few comments apart from that.

avolmat-st commented on dts/arm/st/f4/stm32f469.dtsi:
```diff
@@ -5,6 +5,8 @@
 
 #include <st/f4/stm32f429.dtsi>
+#include <zephyr/dt-bindings/mipi_dsi/mipi_dsi.h>
+
```
Please avoid empty line insertion.

phildefer replied:
fixed

phildefer commented on boards/st/stm32f469i_disco/stm32f469i_disco.dts:
```diff
@@ -63,6 +71,19 @@
 		};
 	};
 
+	dsi_lcd_qsh_030: connector_dsi_lcd {
```
Fully agree, will remove this useless node

phildefer commented on boards/st/stm32f469i_disco/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+#
+# Copyright 2023 BrainCo Inc.
```
updated

phildefer commented on boards/st/stm32f469i_disco/dc_ram.ld:
```diff
@@ -0,0 +1,16 @@
+/*
+ * Copyright (c) 2023 BrainCo Inc.
```
updated

avolmat-st commented on boards/st/stm32f469i_disco/stm32f469i_disco_defconfig:
```diff
@@ -14,3 +14,12 @@ CONFIG_UART_CONSOLE=y
+CONFIG_STM32_LTDC_ARGB8888=y
+
+# Size Memory Heap pool from Framebuffer
+# CONFIG_HEAP_MEM_POOL_SIZE=131072
```
Could you clarify why increasing this MEM_POOL_SIZE is needed ?
I might be wrong but I suspect this is something related to a sample app, and as such this should not be enforced within the board defconfig but instead within the sample code config in order to avoid impacting any application using this board.

I suspect this is what is needed in order to have the samples/drivers/display sample app working since this app is making some heap allocation in order to create the 4 square displayed on each corner of the display.

I've actually a PR #96177 regarding to this where I change the LTDC format and the HEAP_MEM_POOL_SIZE in order to be part of the application side to avoid impacting the board defconfig.
If that is so, could you move both CONFIG_STM32_LTDC_ARGB8888 and CONFIG_HEAP_MEM_POOL_SIZE into the application board specific config file ?

phildefer replied:
@avolmat-st : This line is commented out (sorry i will remove it for clarity) . You are right this framebuffer size is necessary for samples/drivers/display app. I've not digged into details but i've tested yesterday that reducing to 65536 is failing at beginning when running samples/drivers/display apps with this value. This setting is "only" applied when compiling with this samples/drivers/display application (commit https://github.com/zephyrproject-rtos/zephyr/pull/95954/commits/d47ddc4c615e5b271e2f0b567f15014a48d26492) . Not by default in the board .

avolmat-st replied:
Alright, thanks for the explanation. Yes 65K is too small if the framebuffer format is 32bit, 65K would be if the format is 16bit only.
What about having this commit https://github.com/zephyrproject-rtos/zephyr/commit/d47ddc4c615e5b271e2f0b567f15014a48d26492 as part of your PR ? This would allow to have the sample/driver/display sample working how of the box.

avolmat-st reviewed (commented):
Thanks @phildefer for the update.
A few comments added, mainly due to other ongoing PRs which are also impacting this one.
Apart from that, this looks good to me.

avolmat-st commented on boards/st/stm32f469i_disco/Kconfig.defconfig:
```diff
@@ -9,5 +9,36 @@ config SPI_STM32_INTERRUPT
+	default y
+
+choice LV_COLOR_DEPTH
+	default LV_COLOR_DEPTH_16
```
Yeah, sorry about that. While fixing the other points, maybe you might want to put it back to 32 as it was before.

phildefer replied:
revisited, back to 32b COLOR

phildefer commented on boards/st/stm32f469i_disco/dc_ram.ld:
```diff
@@ -0,0 +1,16 @@
+/*
```
done with rebase

phildefer commented on boards/st/stm32f469i_disco/Kconfig.defconfig:
```diff
@@ -9,5 +9,36 @@ config SPI_STM32_INTERRUPT
+config LV_Z_DOUBLE_VDB
+	default y
+
+config LV_Z_VDB_CUSTOM_SECTION
```
done, with rebase

phildefer commented on boards/st/stm32f469i_disco/Kconfig.defconfig:
```diff
@@ -9,5 +9,36 @@ config SPI_STM32_INTERRUPT
+	default y
+
+choice LV_COLOR_DEPTH
+	default LV_COLOR_DEPTH_32
```
revisited as mentioned on other comment

avolmat-st commented on boards/st/stm32f469i_disco/stm32f469i_disco.dts:
```diff
@@ -199,3 +225,59 @@ zephyr_udc0: &usbotg_fs {
 };
+
+&mipi_dsi {
+	status = "okay";
```
Nitpicking (non blocking): prefer putting the status property as the last property within a node. Applicable here but also elsewhere in added code. We try to enforce this more and more now, at least for new added code.

avolmat-st commented on boards/st/stm32f469i_disco/stm32f469i_disco_defconfig:
```diff
@@ -14,3 +14,12 @@ CONFIG_UART_CONSOLE=y
+CONFIG_STM32_LTDC_ARGB8888=y
+
+# Size Memory Heap pool from Framebuffer
+# CONFIG_HEAP_MEM_POOL_SIZE=131072
```
This CONFIG_HEAP_MEM_POOL_SIZE should be removed since there is a board specific conf file in samples/display app.

phildefer replied:
Thanks @avolmat-st , you are right, i will change it

avolmat-st reviewed (changes requested):
Thanks @phildefer for the update.
LGTM, just a very minor point regarding an already commented out line in the defconfig to be removed (MEM_POOL_SIZE) and maybe if you make a new version of the PR you might want as well to correct the ordering in the dts files (status property at the end).
Thanks for your patience.

avolmat-st commented on boards/st/stm32f469i_disco/Kconfig.defconfig:
```diff
@@ -9,5 +9,39 @@ config SPI_STM32_INTERRUPT
+if LVGL
 
+# Double frame buffer maintained by lvgl.
+config STM32_LTDC_FB_NUM
```
Hum, I do feel that this is actually specific to LVGL in the sense that we are saying that when using LVGL we leave the buffer management to LVGL and thus the LTDC driver do not need to allocate a framebuffer since it will come from the LVGL allocated framebuffers.

phildefer commented on dts/arm/st/f4/stm32f469.dtsi:
```diff
@@ -24,5 +25,18 @@
+			clock-names = "dsiclk", "refclk" , "pixelclk";
+			clocks = <&rcc STM32_CLOCK(APB2, 27)>,
+					 <&rcc STM32_SRC_HSE NO_SEL>,
+					 <&rcc STM32_SRC_PLLSAI_DIVR NO_SEL>;
```
updated, as per your recommendations

phildefer commented on boards/st/stm32f469i_disco/Kconfig.defconfig:
```diff
@@ -9,5 +9,39 @@ config SPI_STM32_INTERRUPT
+config LV_Z_FULL_REFRESH
+	default y
+
+config INPUT
```
This part is common with settings present in M1166 display shield

avolmat-st replied:
Having checked a bit more, I am having 2nd thought about this config INPUT in fact.

1. LVGL can work without having an input available (at least it sounds like it)
2. While there is a touchscreen on this display, an application running might not actually want to have the touchscreen initialized / running
3. LVGL samples (at least the ones in samples/modules/lvgl) ALSO set CONFIG_INPUT=y themself.
4. It should also be possible to have INPUT enabled even if there is no LVGL used

So, for all those reasons, I actually think that we just shouldn't have **config INPUT** at all in this Kconfig.defconfig, this is a matter of the project to say if it wants to have the INPUT available or not based on its needs.

Does this make sense ?

While this is also done in several other Kconfig.defconfig around for other STM32 boards/shield, I think this should not be put in here and should actually be removed from the existing places.

erwango replied:
Agreed

avolmat-st replied:
@phildefer, if you agree with my analysis, I let you remove this config INPUT from the Kconfig.default and I will shortly push a PR to avoid enforcing INPUT=y in the STM32 board / shield.

phildefer committed:
samples: subsys: display: add stm32f469_disco conf file

Add stm32f749i_disco conf file to enable input driver
for touchscreen in LVGL samples applications.

Signed-off-by: Philippe Peurichard <p.peurichard@gmail.com>

avolmat-st reviewed (approved):
LGTM. Thanks !!
